File: c++/cote/GJM/특정_구간에서_최소치_두_개_찾기.cpp
#include <iostream>
#include <vector>
#include <random>
#include <limits>
#define TEST_CASES 10
#define N 10000
#define INT_MAX_ numeric_limits<int>::max()

using namespace std; 

struct RMQ {
	int n;
	vector<int*> rangeMin;
	RMQ(const vector<int>& array) {
		n = array.size();
		rangeMin.resize(4 * n);
		init(array, 0, n-1, 1);
	};

	int* query(int left, int right){
		return query(left, right, 1, 0, n-1);
	}

	private : 
		int* init(const vector<int>& array, int left, int right, int node) {
			if (left == right){
				return rangeMin[node] = new int[2] {array[left], INT_MAX_};
			}

			int mid = (left + right) / 2;
			int* leftMin = init(array, left, mid, node * 2);
			int* rightMin = init(array, mid + 1, right, node * 2 + 1);

			vector<int> cand (4);
			copy(leftMin, leftMin+2, begin(cand));
			copy(rightMin, rightMin+2, begin(cand)+2);

			sort(begin(cand), end(cand));

			return rangeMin[node] = new int[]{cand[0], cand[1]};
		}

		int* query(int left, int right, int node, int nodeLeft, int nodeRight) {
			if (right < nodeLeft || nodeRight < left) {
				return new int[2] {INT_MAX_, INT_MAX_};
			}
			if (left <= nodeLeft && nodeRight <= right) {
				return new int[2] {rangeMin[node][0], rangeMin[node][1]};
			}


			int mid = (nodeLeft + nodeRight) / 2;

			int* leftHand = query(left, right, node * 2, nodeLeft, mid);
			int* rightHand = query(left, right, node * 2 + 1, mid + 1, nodeRight);

			vector<int> cand(4);
			copy(leftHand, leftHand+2, begin(cand));
			copy(rightHand, rightHand+2, begin(cand) + 2);

			sort(begin(cand), end(cand));

			delete[] leftHand;
			delete[] rightHand;

			return new int[2]{cand[0], cand[1]};
		}

};


int* solve(int start, int end, vector<int>& v) {
	RMQ rmq(v);
	return rmq.query(start, end);
}

int main(void) {
	default_random_engine gen(43);
	uniform_int_distribution<int> dis(0, N * 5);

	for (int it = 0; it < TEST_CASES; it++) {
		vector<int> v;

		for (int i = 0; i < N; i++){
			v.push_back(dis(gen));
		}

		int end_ = dis(gen) % N + 1;
		int start = dis(gen) % end_;

		sort(begin(v), end(v));
		int* answer = solve(start, end_, v);

		sort(begin(v), end(v));
		cout <<"(" <<answer[0] << ", " << answer[1] << ")" << endl;
		delete[] answer;
	}

	return 0;
}

File: go/cote/GJM/특정_구간에서_최소치_두_개_찾기.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
)

const (
 	TEST_CASES = 10
 	N = 100
 	K = 1000
)

type Range struct{
	left int
	right int
}

type Entry struct{
	first int
	second int
}

func (e Entry) Merge(other Entry) Entry {
	values := []int{0, 0, 0, 0}
	values[0], values[1] = e.first, e.second
	values[2], values[3] = other.first, other.second

	sort.Ints(values)

	return Entry{values[0], values[1]}
}

type RMQ struct {
	n int
	rangeMin []Entry
}

func (rmq *RMQ) Init(array []int) {
	rmq.n = len(array)
	rmq.rangeMin = make([]Entry, 4 * len(array))
	rmq.init(array, 0, len(array), 0)
}

func (rmq RMQ) Query(left, right int) []int {
	result := rmq.query(left, right, 0, rmq.n, 0)
	return []int{result.first, result.second}
}


func (rmq *RMQ) init(array []int, left, right, cur int) Entry {
	if left + 1 == right  {
		rmq.rangeMin[cur] = Entry{array[left], math.MaxInt32}
		return rmq.rangeMin[cur]
	}

	mid := (left + right) / 2
	leftEntry := rmq.init(array, left, mid, 2 * cur + 1)
	rightEntry := rmq.init(array, mid, right, 2 * cur + 2)

	rmq.rangeMin[cur] = leftEntry.Merge(rightEntry)
	return rmq.rangeMin[cur]
}

func (rmq RMQ) query(left, right, curLeft, curRight, cur int) Entry {
	if right <= curLeft || curRight <= left {
		return Entry {math.MaxInt32, math.MaxInt32}
	}

	if left <= curLeft && curRight <= right {
		return rmq.rangeMin[cur]
	}

	mid := (curLeft + curRight) / 2

	leftEntry := rmq.query(left, right, curLeft, mid, 2 * cur + 1)
	rightEntry := rmq.query(left, right, mid, curRight, 2 * cur + 2)

	return leftEntry.Merge(rightEntry)
}

func solve(queries []Range, nums []int) [][]int {

	rmq := new(RMQ)	
	rmq.Init(nums)

	ret := make([][]int, len(queries)) 

	for i, query := range queries {
		ret[i] = rmq.Query(query.left, query.right)
	}

	return ret
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		n := random.Intn(N - 3) + 3
		nums := make([]int, n)
		for i, _ := range nums {
			nums[i] = random.Intn(N * 5)
		}

		k := random.Intn(K - 1) + 1
		queries := make([]Range, k)

		for i, _ := range queries {
			right := random.Intn(n + 1 - 2) + 1 + 2
			left := random.Intn(right - 1)
			queries[i] = Range{left, right}
		}

		fmt.Println(solve(queries, nums))
	}
}